python/whylogs/core/validators/validator.py

```diff
@@ -1,19 +1,42 @@
+import copy
 from abc import ABC, abstractmethod
-from typing import Any, Callable, Dict, List
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, List, Optional, Union
 
 from whylogs.core.metrics.condition_count_metric import Condition
-from whylogs.core.preprocessing import PreprocessedColumn
 
 
+@dataclass
 class Validator(ABC):
     name: str
-    conditions: Dict[str, Condition]
-    actions: List[Callable[[str, str, Any], None]]
+    conditions: Dict[str, Union[Condition, Callable]]
+    actions: List[Union[Callable[[str, str, Any], None], Callable[[str, str, Any, Optional[Any]], None]]]
```

Comment:
Having the relation and actions both on the `Validator` and `Condition` is redundant, but there's probably nothing we can do about it at this point :(

Reply:
agreed, unfortunately